Project 9

Goldshtern opened the pull request.

Goldshtern committed:
Add Api.js class

Goldshtern committed:
Add functions in Api.js class; refactor index.js class

Goldshtern committed:
continue create Api.js class; refactor index.js class

Goldshtern committed:
Add HTML element for
 delete popup

Goldshtern committed:
problem with webpack?

Goldshtern committed:
deleting card is  not perfect

Goldshtern committed:
continue to create Api.js; add DELETE card; reafactor: Card.js, index.js, UserInfo.js

Goldshtern committed:
DELETE method in likes doesnt work at all

Goldshtern committed:
feat: PUT and DELETE methods to likes in Api,js class; feat: addLike, removeLike in index.js class; refactor: Card.js class

Goldshtern committed:
feat: avatar profile udate in HTML and CSS; refactor:profile.css

Goldshtern committed:
feat: add update avatar popUp

Goldshtern committed:
add behavior to the avatar popup; refactor: index.html, UserInfo.js, constants.js, index.js

Goldshtern committed:
connecting website  database via an API, feat: create  Api.js class; refactor: Card.js, Popup.js; Popupwithform.js, index.js, index.html

Goldshtern committed:
fix: index.js

gennady-bars reviewed (commented):
(**Please, expand the general comment ↓**)

Dear Aleksandr, 

unfortunately I have to reject the project without reviewing:



* When I like **new** cards, there are errors https://snipboard.io/45cEQN.jpg
* When I delete 3 **new** cards in a row, there are errors https://snipboard.io/ZPt2il.jpg 
* When I reload the page all new cards, which were deleted, appear again at the bottom of the page

Goldshtern committed:
fix:  handleAddCardFormSubmit(cardData) method in index.js class; fix: renderItems() method in Section.js class

Goldshtern commented:
Gennady, thanks for your feedback. I refactored handleAddCardFormSubmit(cardData) method in index.js class, refactoring constructor in Card.js class. It probably solved problems with delete and likes behavior. I fixed renderItems() method in section.js class, added reverse method to prevent cards from bottom moving.

gennady-bars commented on src/components/Api.js:
```diff
@@ -0,0 +1,108 @@
+  //-----CARD ROUTE: Get all cards (GET)----//
+  getInitialCards() {
+    return fetch(`${this._baseUrl}/cards`, { headers: this._headers })
+      .then(this._checkResponse)
```
**Could be improved**

You can make a special method (function ) for fetching and checking responses **not to duplicate** it in every request:

```js

  _request(url, options) {
    return fetch(url, options).then(this._checkResponse)
  }

```

And now you can safely replace `fetch` with `this._request` and remove repeated checking. And you don’t need to change anything else.

gennady-bars commented on src/components/Api.js:
```diff
@@ -0,0 +1,108 @@
+  getInitialCards() {
+    return fetch(`${this._baseUrl}/cards`, { headers: this._headers })
+      .then(this._checkResponse)
+      .catch((err) => {
```
It’s better not to use `catch` in `Api` because you need to `catch` possible errors at the very **end** of any request and this end is in `index.js`    after all blocks `then`. And in real projects blocks `catch` usually open error popups and so on.

gennady-bars commented on src/components/popupConfirmDelete.js:
```diff
@@ -0,0 +1,17 @@
+    this._popupForm = this._popupElement.querySelector(".modal__form");
+  }
+  setConfirmDelete(handleFormSubmit) {
+    this._handleFormSubmit = handleFormSubmit;
```
Nice confirmation function updating

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -4,6 +4,15 @@ export default class PopupWithForm extends Popup {
+    this._submitButton = this._popupElement.querySelector(".modal__button");
+  }
+
+  viewLoading(isLoading) {
+    if (isLoading) {
+      this._submitButton.textContent = "Saving...";
+    } else {
+      this._submitButton.textContent = "Save";
```
**Could be improved**

It’s better to make method `renderLoading` inside class `PopupWithForm`. You should find the button once in the constructor and change the text:  

```js
  constructor({ popupSelector, handleSubmit }) {
    super(popupSelector);
    // fix the initial button text only once in the constructor
    this._submitBtnText = this._submitBtn.textContent
  }
  // add 2 params: isLoading and loadingText with a default text
  renderLoading(isLoading, loadingText='Saving...') {
    if (isLoading) {
      this._submitBtn.textContent = loadingText;
    } else {
	// here we return back the initial text. So, you don’t need to bother yourself about it
      this._submitBtn.textContent = this._submitBtnText;
    }
  }

```

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -4,6 +4,15 @@ export default class PopupWithForm extends Popup {
     this._handleFormSubmit = handleFormSubmit;
+    this._submitButton = this._popupElement.querySelector(".modal__button");
+  }
+
```
**Could be improved**

You can make ` setInputValues` in class `PopupWithForm`, which can insert data into inputs:

```js

  setInputValues(data) {
    this._inputList.forEach((input) => {
      // Here you insert the `value` by the `name` of the input
      input.value = data[input.name];
    });
  }



```

So, this way you’ll not have to search the inputs of the profile in `index.js`

gennady-bars commented on src/pages/index.js:
```diff
@@ -6,74 +6,134 @@ import PopupWithImage from "../components/PopupWithImage.js";
-  addCardValidator.disableButton();
-  newCardPopup.resetForm();
-  newCardPopup.close();
+  profileEditPopup.viewLoading(true);
```
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```

gennady-bars commented on src/pages/index.js:
```diff
@@ -6,74 +6,134 @@ import PopupWithImage from "../components/PopupWithImage.js";
 //-----------------instantiations------------------------------//
 const profileEditValidator = new FormValidator(config, profileEditForm);
 const addCardValidator = new FormValidator(config, addCardFormElement);
+const addAvatarValidator = new FormValidator(config, editAvatarForm);
```
**Could be improved**

You can **universally** create instances of validators for all forms in the project storing them inside one object: `formValidators`.  Then you can take any validator using attribute `name` of the form where you need to disable/enable the submit button or remove errors.

```js
// define an object for storing validators
const formValidators = {}

const enableValidation = (config) => {
  const formList = Array.from(document.querySelectorAll(config.formSelector))
  formList.forEach((formElement) => {
    const validator = new FormValidator(config, formElement)
    // Here you get the name of the form (if you don’t have it then you need to add it into each form in `index.html` first)
    const formName = formElement.getAttribute('name')

   // Here you store the validator using the `name` of the form
    formValidators[formName] = validator;
    validator.enableValidation();
  });
};

enableValidation(config);

```

And now you can use them for disabling buttons or clearing errors:

```js

formValidators[ profileForm.getAttribute('name') ].resetValidation()

// or you can use a string – the name of the form (you know it from `index.html`)

formValidators['profile-form'].resetValidation()

```

gennady-bars commented on src/pages/index.js:
```diff
@@ -90,4 +150,64 @@ const profileEditPopup = new PopupWithForm(
-const userInfo = new UserInfo(".profile__title", ".profile__description");
+function handleLikeIcon(card) {
+  if (card.isLiked) {
+    api
+      .removeLike(card.id)
+      .then(() => {
+        card.setIsLiked(false);
+      })
+      .catch((err) => console.error(err));
+  } else {
+    api
+      .addLike(card.id)
+      .then(() => {
+        card.setIsLiked(true);
+      })
```
Nice likes handling

gennady-bars commented on src/pages/index.js:
```diff
@@ -90,4 +150,64 @@ const profileEditPopup = new PopupWithForm(
+      .then(() => {
+        card.setIsLiked(false);
+      })
+      .catch((err) => console.error(err));
```
Good job that you `catch` possible errors at the end of server requests. You can simplify it with `.catch(console.error);` because `console.error` is meant to log errors

gennady-bars commented on src/pages/index.js:
```diff
@@ -6,74 +6,134 @@ import PopupWithImage from "../components/PopupWithImage.js";
+      console.err(err);
+    })
+    .finally(() => {
+      avatarEditPopup.viewLoading(false);
```
You need to correctly return the button text of the profile form in `finally`


`profileEditPopup` should be used

gennady-bars commented on src/pages/index.js:
```diff
@@ -6,74 +6,134 @@ import PopupWithImage from "../components/PopupWithImage.js";
+    .then((data) => {
+      userInfo.setUserInfo(data.name, data.about);
+      profileEditValidator.disableButton();
+      profileEditPopup.close();
```
Good job that you **close** popups only in blocks `then`

gennady-bars commented on src/pages/index.js:
```diff
@@ -6,74 +6,134 @@ import PopupWithImage from "../components/PopupWithImage.js";
+      console.err(err);
+    })
+    .finally(() => {
+      newCardPopup.viewLoading(false);
```
Good job that you return the button text in `finally`

gennady-bars reviewed (commented):
**↓Expand to see the whole summary↓**

Hello, Aleksandr!

Great job on submitting your project! This is a significant milestone, and you should be proud of your progress.

**Parts I really liked in your project:**

* Nice confirmation function updating
* Nice likes handling
* Good job that you `catch` possible errors at the end of server requests. 
* Good job that you return the button text in `finally`
* Good job that you **close** popups only in blocks `then`


There are few things that need to be corrected in your project. They're mostly minor issues that are easy to fix.

**Needs correcting:**

* All comments in the code  and the comments below:
* It’s better not to use `catch` in `Api` because you need to `catch` possible errors at the very **end** of any request and this end is in `index.js`    after all blocks `then`. And in real projects blocks `catch` usually open error popups and so on.
* The avatar submit button is active on start though the input is empty https://snipboard.io/zKLbYi.jpg 
* `popupConfirmDelete.js` should be in `UpperCase` like the other files with classes
* You need to correctly return the button text of the profile form in `finally`



There are some comments in your project that can help you make it even better. Please check the `Could be improved` comments.

**A reminder:** It’s necessary to fix all  comments to get the project accepted. And `Could be improved` comments are optional, they will be up to you if you want to implement them.

Your hard work on this project is truly commendable. While there are still some adjustments needed. If you need any assistance with the feedback or understanding certain concepts, our tutors are here to help. Keep moving forward and maintain your dedication — you are doing an amazing job!✨

Goldshtern committed:
refactor, style: Api.js, Card.js, Po
pup.js, PopupWithForm.js, index.js, constants.js, PopupConfirmDelete.js; fix: index.html, index.js

Goldshtern commented:
Hello, Gennady!
      Dear, Gennady, thanks you so much for your feedback! Your advices are very important and useful for me. I tried to implement most of them. I hope I succeeded a little. Sincerely, Aleksandr.

gennady-bars reviewed (approved):
You’ve done a great job!

Your project has been accepted. 

Good luck!
